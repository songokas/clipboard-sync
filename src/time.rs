use chrono::Utc;

#[cfg(feature = "ntp")]
use crate::errors::ConnectionError;
use crate::errors::ValidationError;
#[cfg(feature = "ntp")]
use log::{debug, warn};
#[cfg(feature = "ntp")]
use rsntp::AsyncSntpClient;
#[cfg(feature = "ntp")]
use std::sync::atomic::AtomicI64;

#[cfg(feature = "ntp")]
use std::sync::atomic::Ordering;
use std::time::Duration;

#[cfg(feature = "ntp")]
static TIME_DIFF: AtomicI64 = AtomicI64::new(0);

pub fn get_time() -> u64 {
    #[cfg(feature = "ntp")]
    {
        let diff = TIME_DIFF.load(Ordering::Relaxed);
        let now = Utc::now().timestamp();
        (now + diff) as u64
    }
    #[cfg(not(feature = "ntp"))]
    return Utc::now().timestamp() as u64;
}

pub fn validate_timestamp(
    now: u64,
    timestamp: u64,
    max_expected: Duration,
) -> Result<(), ValidationError> {
    // let now = get_time();
    let difference = if now >= timestamp {
        now - timestamp
    } else {
        timestamp - now
    };
    if max_expected.as_secs() >= difference {
        Ok(())
    } else {
        Err(ValidationError::InvalidTimestamp {
            difference,
            max_expected,
        })
    }
}

#[cfg(feature = "ntp")]
pub async fn update_time_diff(
    ntp_address: String,
    cancel: tokio_util::sync::CancellationToken,
) -> crate::defaults::ExecutorResult {
    use tokio::{
        select,
        time::{self},
    };

    let mut interval = time::interval(Duration::from_secs(1800));
    let mut success_count = 0;
    loop {
        select! {
            _ = cancel.cancelled() => {
                debug!("Ntp cancelled");
                break;
            }
            _ = interval.tick() => {
                let diff = get_time_diff(&ntp_address).await;
                TIME_DIFF.fetch_add(diff, Ordering::Relaxed);
                success_count += 1;
            }
        }
    }
    Ok(("ntp", success_count))
}

#[cfg(feature = "ntp")]
async fn get_time_diff(ntp_address: &str) -> i64 {
    let real = if let Ok(t) = get_ntp_time(ntp_address).await {
        t as i64
    } else {
        return 0;
    };
    let now = Utc::now().timestamp();
    if real >= now {
        real - now
    } else {
        now - real
    }
}

#[cfg(feature = "ntp")]
async fn get_ntp_time(ntp_address: &str) -> Result<u64, ConnectionError> {
    let client = AsyncSntpClient::new();
    let result = client.synchronize(ntp_address).await.map_err(|e| {
        warn!("Failed to synchronize time: {}", e);
        ConnectionError::FailedToConnect(format!(
            "Failed to retrieve time from ntp server {} {}",
            ntp_address, e
        ))
    })?;
    let seconds = result
        .datetime()
        .unix_timestamp()
        .expect("ntp unix timestamp");
    debug!(
        "ntp time updated. server: {} utc time in seconds: {} utc local time: {}",
        ntp_address,
        seconds.as_secs(),
        Utc::now().timestamp()
    );
    Ok(seconds.as_secs())
}

#[cfg(test)]
mod tests {
    use test_data_file::test_data_file;

    use super::*;

    #[test]
    fn test_get_time() {
        assert!(get_time() > 0);
    }

    #[test_data_file(path = "tests/samples/validate_timestamp.csv")]
    #[test]
    fn test_validate_timestamp(now: u64, timestamp: u64, max_valid_secs: u64, is_ok: bool) {
        assert_eq!(
            validate_timestamp(now, timestamp, Duration::from_secs(max_valid_secs)).is_ok(),
            is_ok
        );
    }
}



