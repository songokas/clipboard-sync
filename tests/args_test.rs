use core::time::Duration;

use insta_cmd::assert_cmd_snapshot;

mod common;

use crate::common::run_command;

const ANY_KEY: &str = "12345678912345678912345678912345";

#[test]
fn test_send_once() {
    assert_cmd_snapshot!(run_command(
        vec![
            "--key",
            ANY_KEY,
            "--bind-address",
            "127.0.0.1:8922",
            "--send-once",
            "--clipboard",
            "tests/temp1"
        ],
        Duration::from_secs(2),
        "",
        false,
        vec![],
    ));
}

#[test]
fn test_bind_send_multiple_addresses() {
    assert_cmd_snapshot!(run_command(
        vec![
            "--key",
            ANY_KEY,
            "--bind-address",
            "127.0.0.1:12390,[::1]:12390",
            "--send-using-address",
            "127.0.0.1:12390,[::1]:12390",
            "--send-once",
            "--allowed-host",
            "127.0.0.1:8911,[::1]:8911",
            "--clipboard",
            "tests/temp1"
        ],
        Duration::from_secs(2),
        "",
        true,
        vec![],
    ));
}

#[test]
fn test_send_once_not_immediate() {
    assert_cmd_snapshot!(run_command(
        vec![
            "--key",
            ANY_KEY,
            "--bind-address",
            "127.0.0.1:8923",
            "--send-once",
            "--ignore-initial-clipboard",
        ],
        Duration::from_secs(2),
        "",
        false,
        vec![],
    ));
}

#[test]
fn test_receive_once() {
    assert_cmd_snapshot!(run_command(
        vec![
            "--key",
            ANY_KEY,
            "--bind-address",
            "127.0.0.1:8924",
            "--receive-once",
        ],
        Duration::from_secs(2),
        "",
        true,
        vec![],
    ));
}

#[test]
fn test_config() {
    assert_cmd_snapshot!(run_command(
        vec!["--config", "tests/config.sample.yaml"],
        Duration::from_secs(2),
        "",
        true,
        vec![],
    ));
}

#[test]
fn test_failure_args() {
    for (args, snapshot_name) in get_failure_provider() {
        assert_cmd_snapshot!(
            snapshot_name,
            run_command(args, Duration::from_secs(2), "", false, vec![],)
        );
    }
}

#[test]
fn test_default_with_protocols() {
    for protocol in ["basic", "tcp", "tcp-tls", "quic"] {
        assert_cmd_snapshot!(
            format!("test_default_with_protocols_{protocol}"),
            run_command(
                vec![
                    "--protocol",
                    protocol,
                    "--key",
                    ANY_KEY,
                    "--bind-address",
                    "[::]:12001",
                    "--allowed-host",
                    "[::1]:0",
                    "--ignore-initial-clipboard"
                ],
                Duration::from_secs(2),
                "",
                false,
                vec![],
            )
        );
    }
}

fn get_failure_provider() -> Vec<(Vec<&'static str>, &'static str)> {
    vec![
        (vec!["--key", "3232"], "invalid_key"),
        (vec!["--config", "_1_unknow_path"], "unknown_config_path"),
        (
            vec!["--key", ANY_KEY, "--unknown", "other"],
            "unexpected argument",
        ),
        (
            vec!["--key", ANY_KEY, "--send-using-address", "non-existing"],
            "non-existing-send-address",
        ),
        (
            vec!["--key", ANY_KEY, "--bind-address", "non-existing"],
            "non-existing-bind-address",
        ),
        (
            vec!["--key", ANY_KEY, "--protocol", "non-existing"],
            "non-existing-protocol",
        ),
    ]
}
