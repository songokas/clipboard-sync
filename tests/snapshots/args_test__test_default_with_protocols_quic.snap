---
source: tests/args_test.rs
info:
  program: clipboard-sync
  args:
    - "--protocol"
    - quic
    - "--key"
    - "12345678912345678912345678912345"
    - "--bind-address"
    - "[::]:12001"
    - "--allowed-host"
    - "[::1]:0"
    - "--ignore-initial-clipboard"
    - "--remote-certificates"
    - tests/certs/cert-verify/for-server
    - "--private-key"
    - tests/certs/localhost.key
    - "--certificate-chain"
    - tests/certs/localhost.crt
    - "--verbosity"
    - info=simple
snapshot_kind: text
---
success: false
exit_code: -1
----- stdout -----

----- stderr -----
Listening on local_addr=[::]:12001 protocol=quic multicast_addr=None certificate_serial=46:8d:f7:de:e1:27:f9:0d:06:c6:1f:ea:05:36:d8:2e:ad:69:09:35 certificate_dns=localhost
